addons/product/models/product_template.py

```diff
@@ -502,6 +543,226 @@ def has_dynamic_attributes(self):
+           exclusions.exclusions: exclusions within this product
+           exclusions.parent_exclusions: exclusions coming from the reference_product
+        """
+        self.ensure_one()
+
+        product_attribute_values = self.env['product.template.attribute.value'].search([
+            ('product_tmpl_id', '=', self.id),
+            ('product_attribute_value_id', 'in', self.attribute_line_ids.mapped('value_ids').ids),
+        ])
+
+        # array of all the excluded value_ids of all the filter lines for this product
+        mapped_exclusions = {
+            product_attribute_value.id: [
+                value_id
+                for filter_line in product_attribute_value.exclude_for.filtered(
+                    lambda filter_line: filter_line.product_tmpl_id == self
+                ) for value_id in filter_line.value_ids.ids
+            ]
+            for product_attribute_value in product_attribute_values
+        }
+
+        parent_exclusions = []
+        if reference_product:
+            parent_attribute_value_ids = reference_product.product_template_attribute_value_ids
+            if parent_attribute_value_ids and reference_product._context.get('no_variant_attribute_values'):
+                # Add "no_variant" attribute values' exclusions
+                # They are kept in the context since they are not linked to this product variant
+                parent_attribute_value_ids |= reference_product._context.get('no_variant_attribute_values')
+
+            parent_exclusions = [
+                value_id
+                for filter_line in parent_attribute_value_ids.mapped('exclude_for').filtered(
+                    lambda filter_line: filter_line.product_tmpl_id == self
+                ) for value_id in filter_line.value_ids.ids]
+
+        # Query all archived products for this template
+        archived_combinations = self.env['product.product'].search(
+            [('product_tmpl_id', '=', self.id), ('active', '=', False)])
+
+        if archived_combinations:
+            # Old archived variants could have a different set of attributes and are not relevant here
+            # -> filter them out
+            attribute_ids = product_attribute_values.mapped('attribute_id')
+            archived_combinations = archived_combinations.filtered(
+                lambda product: all(
+                    attribute_id in product.mapped('product_template_attribute_value_ids.attribute_id')
+                    for attribute_id in attribute_ids
+                )
+            )
+
+        return {
+            'exclusions': mapped_exclusions,
+            'parent_exclusions': parent_exclusions,
+            'archived_combinations': [archived_combination.product_template_attribute_value_ids.ids
+                for archived_combination in archived_combinations]
+        }
+
+    @api.model
+    def _get_combination_info(self, product_template_id, product_id, combination, add_qty, pricelist):
```

Comment:
@seb-odoo 

If this is called on the product_template model, it probably shouldn't receive "product_template_id" as an argument.
The caller could do "env['product.template'].browse(product_template_id)._get_combination_info(product_id, combination, add_qty, pricelist)

Reply:
I was wondering the same, but then what to do with the special context? In the code you made, you replaced the context before doing the browse. Is it important that the context be changed before the browse, or can it be done after.

Reply:
It's not important, you can replace the browse line with "self = self.with_context()" I think.

You can easily test it after modifying the code by setting a price_extra on dynamically created products and see if the price correctly changes in the UI when the product is not created yet.

Reply:
I changed this as suggested.